### PR TITLE
Adds color support for wrapped tables

### PR DESCRIPTION
**Pull request by cgranleese-r7**
This PR adds color support for wrapped tables, as well as tests to cover as many edge cases as possible.

## Before
image

## After
image

## Prerequisites
Checkout this branch and add the `rex-text` gem into your Metasploit `gemfile`.
```Ruby
gemspec name: 'metasploit-framework'
gem "rex-text", path: "../rex-text"
```
Set [WordWrap](https://github.com/rapid7/metasploit-framework/blob/5d935bdd0a25ff20d01f67e3c822685fc1c79aee/lib/msf/ui/console/command_dispatcher/modules.rb#L16) to `true` within `lib/msf/ui/console/command_dispatcher/modules.rb`

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] Verify `search` command highlight expected result(s)
- [ ] Adjust terminal window size
- [ ] Verify again that `search` command highlight expected result(s)
- [ ]  Verify table are being wrapped as expected

### CODE REVIEW

File: lib/rex/text/wrapped_table.rb [hunk 1]
```diff
@@ -401,17 +402,104 @@ def row_to_s(row) # :nodoc:
...
 
+  #
+  # Takes a hash of color/format tags and converts them to a string
+  #
+  # @param [Hash] color_state
+  # @param [Integer] char_index
+  def color_code_string_for(color_state, char_index)
+    color_str = ""
+
+    unless char_index == 0
```
Comment (dwelch-r7) [inline lib/rex/text/wrapped_table.rb:416]: Might be more readable to return early if `char_index` is 0 rather than nesting in an `unless`

File: lib/rex/text/wrapped_table.rb [hunk 2]
```diff
@@ -401,17 +402,104 @@ def row_to_s(row) # :nodoc:
...
+      # Check if any color code(s) from previous the string need appended
+      while char_index < chars.length do
+        if current_chunk.nil?
+          previous_color_codes = color_code_string_for(color_state, char_index)
+          current_chunk = previous_color_codes
+        end
+
+        # Matches color codes and adds them to the color_state hash, uses color/formats purpose as key
+        # e.g. {:foreground=>"%blu"}
+        if chars[char_index] == '%'
```
Comment (dwelch-r7) [inline lib/rex/text/wrapped_table.rb:457]: have you tested what happens if you have a `%` symbol in the data _not_ part of a formatting code? might be missing it but looks like you might not advanced `char_index` in that case

File: lib/rex/text/wrapped_table.rb [hunk 3]
```diff
@@ -401,17 +402,104 @@ def row_to_s(row) # :nodoc:
   # widths. For now it simply returns the string's length.
   #
   def display_width(str)
-    str.length
+    Rex::Text.display_width(str)
   end
 
+  #
+  # Takes a hash of color/format tags and converts them to a string
```
Comment (dwelch-r7) [inline lib/rex/text/wrapped_table.rb:409]: I don't think this comment really explains what's going on here
what's the purpose of converting them to a string? what does the resulting string look like?

File: lib/rex/text/wrapped_table.rb (hunk 2 above)
Comment (cgranleese-r7) [reply lib/rex/text/wrapped_table.rb:457]: Good catch, I'll get a test added to cover that 👍

File: lib/rex/text/wrapped_table.rb [hunk 4]
```diff
@@ -401,17 +402,104 @@ def row_to_s(row) # :nodoc:
...
+            color_state[:format] = []
+            color_state[:format] << [value[char_index..(char_index + 3)]]
+            current_chunk << value[char_index..(char_index + 3)]
+            char_index += 3
+          elsif value[char_index..(char_index + 3)] == '%clr'
+            color_state.clear
+            current_chunk << '%clr'
+            char_index += 3
+          elsif Rex::Text::Color::SUPPORTED_FORMAT_CODES.include?(value[char_index..(char_index + 5)])
+            color_state[:background] = value[char_index..(char_index + 5)] unless value[char_index..(char_index + 5)].end_with?('%clr')
```
[the remark this answers is not in the document]
Comment (cgranleese-r7) [reply lib/rex/text/wrapped_table.rb:468]: This was to cover it matching on a `%clr` at the end of the string. It was still matching even though it was technically not the full range.

File: lib/rex/text/wrapped_table.rb [hunk 5]
```diff
@@ -401,17 +402,104 @@ def row_to_s(row) # :nodoc:
...
+            current_chunk << value[char_index..(char_index + 5)]
+            char_index += 5
+          elsif Rex::Text::Color::SUPPORTED_FORMAT_CODES.include?(value[char_index..(char_index + 3)])
+            color_state[:foreground] = value[char_index..(char_index + 3)]
+            current_chunk << value[char_index..(char_index + 3)]
+            char_index += 3
+          end
+        else
+          current_chunk << chars[char_index]
```
[the remark this answers is not in the document]
Comment (cgranleese-r7) [reply lib/rex/text/wrapped_table.rb:477]: These increments were added to skip the remainder of any color/formatting that was matched and added to the `color_state` hash.

e.g. if we match on `%bgyel`, rather than looping over the 5 trailing characters to check if they match `%`, I just increased the index to skip those characters.

File: lib/rex/text/wrapped_table.rb [hunk 6]
```diff
@@ -401,17 +402,104 @@ def row_to_s(row) # :nodoc:
...
+
+        # display_width will calculate the string's length ignoring color codes
+        if display_width(current_chunk) == column_width
+          unless color_state.empty?
+            current_chunk.insert(-1, '%clr') unless current_chunk.end_with?('%clr')
+          end
+          chunks.push(current_chunk)
+          current_chunk = nil
+        end
+        char_index += 1
```
[the remark this answers is not in the document]
Comment (cgranleese-r7) [reply lib/rex/text/wrapped_table.rb:488]: My increments within the loop were done so increment to the last character of the string to skip, then after the logic I'd increment by 1 regardless, if that makes sense.

image


But I think moving it up regardless makes more sense anyway for readability

File: lib/rex/text/wrapped_table.rb (hunk 1 above)
Comment (cgranleese-r7) [reply lib/rex/text/wrapped_table.rb:418]: Yea looking at it again, would be better to just return early 👍

`char_index` is being passed so it won't add return the `color_str` if we move to another table column. So I check if `char_index` is 0 to verify if we're at the beginning of a column.

e.g. we don't want any of the `color_state` data from the column 1 to be added to column 2.

File: lib/rex/text/wrapped_table.rb [hunk 7]
```diff
@@ -401,17 +402,104 @@ def row_to_s(row) # :nodoc:
...
   def display_width(str)
-    str.length
+    Rex::Text.display_width(str)
   end
 
+  #
+  # Takes a hash of color/format tags and converts them to a string
+  #
+  # @param [Hash] color_state
+  # @param [Integer] char_index
```
[the remark this answers is not in the document]
Comment (cgranleese-r7) [reply lib/rex/text/wrapped_table.rb]: Updated the comment to hopefully add more clarity.

File: lib/rex/text/wrapped_table.rb [hunk 8]
```diff
@@ -401,17 +402,107 @@ def row_to_s(row) # :nodoc:
...
 
+  #
+  # Returns a string of color/formatting codes made up of the previously stored color_state
+  # e.g. if a `%blu` color code spans multiple lines this will return a string of `%blu` to be appended to
+  # the beginning of each row
+  #
+  # @param color_state [Hash] tracks current color/formatting codes within table row
+  # @param char_index [Integer] tracks current character index across table rows
+  def color_code_string_for(color_state, char_index)
+    color_str = ""
```
Comment (dwelch-r7) [inline lib/rex/text/wrapped_table.rb:416]:
```suggestion
    color_str = ''
```

File: lib/rex/text/wrapped_table.rb [hunk 9]
```diff
@@ -401,17 +402,107 @@ def row_to_s(row) # :nodoc:
...
+    return color_str if char_index == 0
```
Comment (dwelch-r7) [inline lib/rex/text/wrapped_table.rb:417]: this seems weird to me, the only reason you pass in char_index is to short circuit if it's 0, I feel like that sort of falls out of the responsibility of this function and it just shouldn't be called if char_index is 0?

File: lib/rex/text/wrapped_table.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (dwelch-r7) [reply lib/rex/text/wrapped_table.rb:469]: do you have an example or test that hits this?

File: lib/rex/text/wrapped_table.rb (hunk 9 above)
Comment (cgranleese-r7) [reply lib/rex/text/wrapped_table.rb:417]: Good point, I'll make that conditional call outside of the method 👍

File: lib/rex/text/wrapped_table.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (cgranleese-r7) [reply lib/rex/text/wrapped_table.rb]: Yea after looking into this more, this was left over from one of my earlier implementations and is no longer required due to `%clr` now being match within that logic.

File: lib/rex/text/wrapped_table.rb [hunk 10]
```diff
@@ -519,13 +610,12 @@ def format_table_field(str, idx)
     str_cp
   end
 
-  def style_table_field(str, _idx)
-    str_cp = str.dup
+  def style_table_field(str, idx)
+    str_cp = str.clone
```
Comment (adfoster-r7) [inline lib/rex/text/wrapped_table.rb:614]: This should be `dup` still - not sure why this was swapped over 👀 
https://www.rubyguides.com/2018/11/dup-vs-clone/

Comment (adfoster-r7) [reply lib/rex/text/wrapped_table.rb:614]: Ah, maybe this was accidental accidental copy/pasta from `table.rb` when you were debugging things

File: spec/rex/text/wrapped_table_spec.rb [hunk 11]
```diff
@@ -609,17 +748,213 @@ def style(str)
...
+          -----------                             ----------
+          %blu%undA%magA%yelAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%clr  %red%undA%magA%yelAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%clr
+          %yel%undAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%clr  %yel%undAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%clr
+          %yel%undAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%clr     %yel%undAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%clr
+        TABLE
+
+        expect(tbl.to_s.lines).to all(have_maximum_display_width(80))
+      end
+
+      it "verify that no formatting is added when not required" do
```
Comment (adfoster-r7) [inline spec/rex/text/wrapped_table_spec.rb:830]: Some of these descriptions seem to missing english words 👀

File: spec/rex/text/wrapped_table_spec.rb [hunk 12]
```diff
@@ -585,18 +696,46 @@ def style(str)
...
-        expect(tbl.to_s.lines).to all(have_maximum_width(80))
+        expect(tbl.to_s.lines).to all(have_maximum_display_width(80))
       end
 
-      it "safely wordwraps cells that have colors present" do
-        skip(
-          "Functionality not implemented. Currently if there are colors present in a cell, the colors will break " \
-          "when word wrapping occurs"
-        )
+      it "safely wordwraps cells when a % not associated with a color/format codes" do
```
Comment (adfoster-r7) [inline spec/rex/text/wrapped_table_spec.rb:702]: Some of these descriptions seem to missing english words 👀

File: lib/rex/text/wrapped_table.rb [hunk 13]
```diff
@@ -401,17 +402,106 @@ def row_to_s(row) # :nodoc:
...
+          # color_color_codes to a blank string and not carry over previous color codes
+          char_index == 0 ? previous_color_codes = '' : previous_color_codes = color_code_string_for(color_state)
+          current_chunk = previous_color_codes
+        end
+
+        # Matches color codes and adds them to the color_state hash, uses color/formats purpose as key
+        # e.g. {:foreground=>"%blu"}
+        if chars[char_index] == '%'
+          if value[char_index..(char_index + 3)] == '%und' ||  value[char_index..(char_index + 3)] == '%bld'
+            color_state[:format] = []
```
Comment (adfoster-r7) [inline lib/rex/text/wrapped_table.rb:459]: This looks like a bug, I assume this is meant to be `||=`? 🤔 

As something can be both bold and underlined at the same time

File: lib/rex/text/wrapped_table.rb [hunk 14]
```diff
@@ -401,17 +402,106 @@ def row_to_s(row) # :nodoc:
...
+          # char_index == 0 signals the beginning of a new column, therefore we want to set the previous
+          # color_color_codes to a blank string and not carry over previous color codes
+          char_index == 0 ? previous_color_codes = '' : previous_color_codes = color_code_string_for(color_state)
+          current_chunk = previous_color_codes
+        end
+
+        # Matches color codes and adds them to the color_state hash, uses color/formats purpose as key
+        # e.g. {:foreground=>"%blu"}
+        if chars[char_index] == '%'
+          if value[char_index..(char_index + 3)] == '%und' ||  value[char_index..(char_index + 3)] == '%bld'
```
Comment (adfoster-r7) [inline lib/rex/text/wrapped_table.rb:458]: I believe this is wrong here; `char_index` matches to an index in the characters array, not the original value string. i.e. `chars.length != value.length`

File: lib/rex/text/wrapped_table.rb [hunk 15]
```diff
@@ -401,17 +402,106 @@ def row_to_s(row) # :nodoc:
...
+      current_chunk = nil
+      chars = value.chars
+      char_index = 0
+
+      # Check if any color code(s) from previous the string need appended
+      while char_index < chars.length do
+        if current_chunk.nil?
+          # char_index == 0 signals the beginning of a new column, therefore we want to set the previous
+          # color_color_codes to a blank string and not carry over previous color codes
+          char_index == 0 ? previous_color_codes = '' : previous_color_codes = color_code_string_for(color_state)
```
Comment (adfoster-r7) [inline lib/rex/text/wrapped_table.rb:451]: It's best practice to avoid using ternary operators to cause side effects, such as assignments

Normally you would write this like:

```
previous_color_codes = char_index == 0 ? '' : color_code_string_for(color_state)
```

Whilst currently you're using a ternary operator almost like an if statement:
```
if char_index == 0
  previous_color_codes = ''
else
  previous_color_codes = color_code_string_for(color_state)
end
```

**Comment by adfoster-r7** [comment]
Here's a diff to fix up some of the issues I spotted when code reviewing, and I removed some code duplication. I'm sure it also introduces new bugs, but the existing tests and new tests pass, as well as fixing one previously incorrect test:

```diff
diff --git a/lib/rex/text/wrapped_table.rb b/lib/rex/text/wrapped_table.rb
index f9cf45a..05f4b48 100644
--- a/lib/rex/text/wrapped_table.rb
+++ b/lib/rex/text/wrapped_table.rb
@@ -410,17 +410,59 @@ protected
   # e.g. if a `%blu` color code spans multiple lines this will return a string of `%blu` to be appended to
   # the beginning of each row
   #
-  # @param color_state [Hash<String, String>] tracks current color/formatting codes within table row
+  # @param [Hash<String, String>] color_state tracks current color/formatting codes within table row
+  # @returns [String] Color code string such as `%blu%grn'
   def color_code_string_for(color_state)
-    color_str = ''
+    result = ''.dup
     color_state.each do |_format, value|
       if value.is_a?(Array)
-        color_str += value.join
+        result << value.uniq.join
       else
-        color_str += value
+        result << value
       end
     end
-    color_str
+    result
+  end
+
+  # @returns [Hash<String, String>] The supported color codes from {Rex::Text::Color} grouped into sections
+  def color_code_groups
+    return @color_code_groups if @color_code_groups
+
+    @color_code_groups = {
+      foreground: %w[
+        %cya %red %grn %blu %yel %whi %mag %blk
+        %dred %dgrn %dblu %dyel %dcya %dwhi %dmag
+      ],
+      background: %w[
+        %bgblu %bgyel %bggrn %bgmag %bgblk %bgred %bgcyn %bgwhi
+      ],
+      decoration: %w[
+        %und %bld
+      ],
+      clear: %w[
+        %clr
+      ]
+    }
+
+    # Developer exception raised to ensure all color codes are accounted for. Verified via tests.
+    missing_color_codes = (Rex::Text::Color::SUPPORTED_FORMAT_CODES - @color_code_groups.values.flatten)
+    raise "Unsupported color codes #{missing_color_codes.join(', ')}" if missing_color_codes.any?
+
+    @color_code_groups
+  end
+
+  # Find the preceding color type and value of a given string
+  # @param [String] string A string such as '%bgyel etc'
+  # @returns [Array,nil] A tuple with the color type and value, or nil
+  def find_color_type_and_value(string)
+    color_code_groups.each do |color_type, color_values|
+      color_value = color_values.find { |color_value| string.start_with?(color_value) }
+      if color_value
+        return [color_type, color_value]
+      end
+    end
+
+    nil
   end

   #
@@ -433,10 +475,9 @@ protected
   # @param [Array<String>] values
   # @param [Integer] optimal_widths
   def chunk_values(values, optimal_widths)
-    color_state = {}
-
     # First split long strings into an array of chunks, where each chunk size is the calculated column width
     values_as_chunks = values.each_with_index.map do |value, idx|
+      color_state = {}
       column_width = optimal_widths[idx]
       chunks = []
       current_chunk = nil
@@ -445,62 +486,49 @@ protected

       # Check if any color code(s) from previous the string need appended
       while char_index < chars.length do
-        if current_chunk.nil?
-          # char_index == 0 signals the beginning of a new column, therefore we want to set the previous
-          # color_color_codes to a blank string and not carry over previous color codes
-          char_index == 0 ? previous_color_codes = '' : previous_color_codes = color_code_string_for(color_state)
-          current_chunk = previous_color_codes
+        # If a new chunk has started, start the chunk with any previous color codes
+        if current_chunk.nil? && color_state.any?
+          current_chunk = color_code_string_for(color_state)
         end
+        current_chunk ||= ''.dup
+
+        # Check if the remaining chars start with a color code such as %blu
+        color_type_and_value = chars[char_index] == '%' ? find_color_type_and_value(chars[char_index..].join) : nil
+        if color_type_and_value.nil?
+          current_chunk << chars[char_index]
+          char_index += 1
+        else
+          color_type, color_code = color_type_and_value

-        # Matches color codes and adds them to the color_state hash, uses color/formats purpose as key
-        # e.g. {:foreground=>"%blu"}
-        if chars[char_index] == '%'
-          if value[char_index..(char_index + 3)] == '%und' ||  value[char_index..(char_index + 3)] == '%bld'
-            color_state[:format] = []
-            color_state[:format] << [value[char_index..(char_index + 3)]]
-            current_chunk << value[char_index..(char_index + 3)]
-            char_index += 3
-          elsif value[char_index..(char_index + 3)] == '%clr'
+          if color_type == :clear
             color_state.clear
-            current_chunk << '%clr'
-            char_index += 3
-          elsif Rex::Text::Color::SUPPORTED_FORMAT_CODES.include?(value[char_index..(char_index + 5)])
-            color_state[:background] = value[char_index..(char_index + 5)]
-            current_chunk << value[char_index..(char_index + 5)]
-            char_index += 5
-          elsif Rex::Text::Color::SUPPORTED_FORMAT_CODES.include?(value[char_index..(char_index + 3)])
-            color_state[:foreground] = value[char_index..(char_index + 3)]
-            current_chunk << value[char_index..(char_index + 3)]
-            char_index += 3
+          elsif color_type == :decoration
+            # Multiple decorations can be enabled
+            color_state[:decoration] ||= []
+            color_state[:decoration] << color_code
           else
-            current_chunk << chars[char_index]
+            # There can only be one foreground or background color
+            color_state[color_type] = color_code
           end
-        else
-          current_chunk << chars[char_index]
+
+          current_chunk << color_code
+          char_index += color_code.length
         end
-        char_index += 1

-        # display_width will calculate the string's length ignoring color codes
-        if display_width(current_chunk) == column_width
-          unless color_state.empty?
-            current_chunk.insert(-1, '%clr') unless current_chunk.end_with?('%clr')
+        # If we've reached the final character of the string, or need to word wrap
+        # it's time to push the current chunk, and start a new row. Also discard
+        # any values that are purely colors and have no display_width
+        is_final_character = char_index >= chars.length
+        display_width = display_width(current_chunk)
+        if (is_final_character && display_width != 0) || display_width == column_width
+          if color_state.any? && !current_chunk.end_with?('%clr')
+            current_chunk << '%clr'
           end
           chunks.push(current_chunk)
           current_chunk = nil
         end
       end

-      if current_chunk != nil
-        unless color_state.empty?
-          current_chunk.insert(-1, '%clr') unless current_chunk.end_with?('%clr')
-        end
-
-        unless display_width(current_chunk) == 0
-          chunks.push(current_chunk)
-        end
-      end
-
-      color_state.clear
       chunks
     end

diff --git a/spec/rex/text/wrapped_table_spec.rb b/spec/rex/text/wrapped_table_spec.rb
index 4460a5a..d0dc77f 100644
--- a/spec/rex/text/wrapped_table_spec.rb
+++ b/spec/rex/text/wrapped_table_spec.rb
@@ -879,7 +879,7 @@ describe Rex::Text::Table do
         expect(tbl).to match_table <<~TABLE
           Header
           ======
-
+
           Blue Column                             Red Column
           -----------                             ----------
           %bluAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%clr  %redAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%clr
@@ -911,7 +911,7 @@ describe Rex::Text::Table do
         expect(tbl).to match_table <<~TABLE
           Header
           ======
-
+
           Blue Column                             Red Column
           -----------                             ----------
           %bluAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%clr  %redAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%clr
@@ -921,6 +921,36 @@ describe Rex::Text::Table do
         expect(tbl.to_s.lines).to all(have_maximum_display_width(80))
       end

+      it "supports all dark color codes" do
+
+        options = {
+          'Header' => 'Header',
+          'Indent' => 0,
+          'Width' => 80,
+          'Columns' => [
+            'Blue Column',
+            'Red Column'
+          ]
+        }
+
+        tbl = Rex::Text::Table.new(options)
+        tbl << [
+          "%dyel#{'A' * 40}%clr",
+          "%dcya#{'A' * 40}%clr",
+        ]
+
+        expect(tbl).to match_table <<~TABLE
+          Header
+          ======
+
+          Blue Column                             Red Column
+          -----------                             ----------
+          %dyelAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%clr  %dcyaAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%clr
+          %dyelAA%clr                                      %dcyaAA%clr
+        TABLE
+
+        expect(tbl.to_s.lines).to all(have_maximum_display_width(80))
+      end

       it "verify that formatting aligns correctly" do

@@ -950,8 +980,8 @@ describe Rex::Text::Table do
           n
           -  -
           %bgyel%blu%bld%undA%clr  %bgyel%blu%bld%undA%clr
-          %bgyel%blu%undA%clr  %bgyel%blu%undA%clr
-          %bgyel%blu%undA%clr  %bgyel%blu%undA%clr
+          %bgyel%blu%bld%undA%clr  %bgyel%blu%bld%undA%clr
+          %bgyel%blu%bld%undA%clr  %bgyel%blu%bld%undA%clr
         TABLE

         expect(tbl.to_s.lines).to all(have_maximum_display_width(7))
```

File: lib/rex/text/wrapped_table.rb (hunk 10 above)
Comment (cgranleese-r7) [reply lib/rex/text/wrapped_table.rb:614]: Yea, that's exactly what happened 😅

File: lib/rex/text/wrapped_table.rb (hunk 13 above)
Comment (cgranleese-r7) [reply lib/rex/text/wrapped_table.rb:459]: Ah, good catch 👍

**Comment by adfoster-r7** [comment]
Working as expected

Before

image

After

image
